rama-net/src/address/socket_address.rs

```diff
@@ -0,0 +1,257 @@
+        match &self.ip_addr {
+            IpAddr::V4(ip) => write!(f, "{}:{}", ip, self.port),
+            IpAddr::V6(ip) => write!(f, "[{}]:{}", ip, self.port),
+        }
+    }
+}
+
+impl FromStr for SocketAddress {
+    type Err = OpaqueError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        SocketAddress::try_from(s)
+    }
+}
+
+impl TryFrom<String> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        s.as_str().try_into()
+    }
+}
+
+impl TryFrom<&str> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        let (ip_addr, port) = crate::address::authority::split_port_from_str(s)?;
+        let ip_addr = IpAddr::from_str(ip_addr).context("parse ip address from socket address")?;
+        match ip_addr {
+            IpAddr::V6(_) if !s.starts_with('[') => Err(OpaqueError::from_display(
+                "missing brackets for IPv6 address with port",
+            )),
+            _ => Ok(SocketAddress { ip_addr, port }),
+        }
+    }
+}
+
+#[cfg(feature = "http")]
+impl TryFrom<HeaderValue> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(header: HeaderValue) -> Result<Self, Self::Error> {
+        Self::try_from(&header)
+    }
+}
+
+#[cfg(feature = "http")]
+impl TryFrom<&HeaderValue> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(header: &HeaderValue) -> Result<Self, Self::Error> {
+        header.to_str().context("convert header to str")?.try_into()
+    }
+}
+
+impl TryFrom<Vec<u8>> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(bytes.as_slice())
+    }
+}
+
+impl TryFrom<&[u8]> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let s = std::str::from_utf8(bytes).context("parse authority from bytes")?;
+        s.try_into()
+    }
+}
+
+impl serde::Serialize for SocketAddress {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let address = self.to_string();
+        address.serialize(serializer)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SocketAddress {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        s.try_into().map_err(serde::de::Error::custom)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn assert_eq(s: &str, sock_address: SocketAddress, ip_addr: &str, port: u16) {
+        assert_eq!(
+            sock_address.ip_addr().to_string(),
+            ip_addr,
+            "parsing: {}",
+            s
+        );
+        assert_eq!(sock_address.port(), port, "parsing: {}", s);
+    }
+
+    #[test]
+    fn test_parse_valid() {
```

Comment:
You copied this from Authority I think. Where I see the same `fn test_parse_valid() {`.

But looking at it, I think that test doesn't do anything useful. This is not your fault as my original test is similar. But now that you're working on this it might be good to verify and if needed fix it.

A line like:

```
assert_eq(s, s.parse().expect(&msg), expected_ip_addr, expected_port)
```

The first is a string (`s`) and the second is `s.parse().expect(...)`, meaning the second one parses into the same type as the first, ... which is a string.

Probably need like a valid actual SocketAddr as a new added first parameter, so that `parse()` actually deserialized into a SocketAddress. Can you verify this and if needed fix it for both `authority.rs` and `socket_address.rs`.